src/DevBetterWeb.Core/ValueObjects/BillingDetails.cs

```diff
@@ -1,24 +1,115 @@
+﻿using System.Collections.Generic;
 using CSharpFunctionalExtensions;
+using DevBetterWeb.Core.Enums;
+using System;
 
 namespace DevBetterWeb.Core.ValueObjects
 {
   public class BillingDetails : ValueObject
   {
-    public string Message { get; }
     public decimal Amount { get; }
+    public string MemberName { get; }
+    public string SubscriptionPlanName { get; }
+    public string ActionVerbPastTense { get; }
+    public BillingPeriod BillingPeriod { get; }
+    public DateTime Date { get; }
 
-    public BillingDetails(string message, decimal amount = 0)
+    public BillingDetails(string memberName, string subscriptionPlanName, string actionVerbPastTense, BillingPeriod billingPeriod, DateTime date, decimal amount = 0)
     {
-      Message = message;
       Amount = amount;
+      MemberName = memberName;
+      SubscriptionPlanName = subscriptionPlanName;
+      ActionVerbPastTense = actionVerbPastTense;
+      BillingPeriod = billingPeriod;
+      Date = date;
+    }
+
+    public string GetMessageForMemberView()
+    {
+      string message = GetSentenceWithProperSyntax(Viewer.Member);
+
+      return message;
+    }
+
+    public string GetMessageForAdminView(string memberId)
+    {
+      string message = GetSentenceWithProperSyntax(Viewer.Admin, memberId);
+      return message;
+    }
+
+    private string GetSentenceWithProperSyntax(Viewer viewer, string memberId = "")
+    {
+      string message = "";
+
+      switch(viewer)
+      {
+        case Viewer.Admin:
+          message += $"Member ID: {memberId}. {MemberName}";
+          break;
+        case Viewer.Member:
+          message += $"You";
+          break;
+        default:
+          break;
+      }
+
+      switch (ActionVerbPastTense)
+      {
+        case "Subscribed":
+          message += $" {ActionVerbPastTense} to {SubscriptionPlanName} for ${Amount}";
+          break;
+        case "Ended":
+          message += $"{GetPossessiveEnding(viewer)} {SubscriptionPlanName} {ActionVerbPastTense}";
+          break;
+        case "Renewed":
+          message += $" {ActionVerbPastTense} {SubscriptionPlanName} for ${Amount}";
+          break;
+        case "Cancelled":
+        default:
+          message += $" {ActionVerbPastTense} {SubscriptionPlanName}";
+          break;
+      }
+
+      message += $" on {Date.ToLongDateString()}.";
+
+      return message;
+    }
+
+    private string GetPossessiveEnding(Viewer viewer)
+    {
+      switch(viewer)
+      {
+        case Viewer.Admin:
+          return "'s";
+        case Viewer.Member:
+          return "r";
+        default:
+          return "";
+      }
     }
 
     protected override IEnumerable<object> GetEqualityComponents()
     {
       yield return Amount;
-      yield return Message;
+      yield return SubscriptionPlanName;
+      yield return MemberName;
+      yield return ActionVerbPastTense;
+      yield return BillingPeriod;
+    }
+
+    public BillingDetails()
```

Comment:
No need to add constructor initialization for default values, and whenever possible prefer setting fields/properties inline (where they are declared) rather than in the constructor.

Save the constructor for:
1) assigning arguments passed into it
2) validation/guard class on such arguments
3) initialization that can't be done inline or as the default for the type

Reply:
I need this for EF Core, don't I? Since the properties aren't nullable some default needs to be given